cmd/mt-parrot/monitor.go

```diff
@@ -0,0 +1,154 @@
+
+type seriesStats struct {
+	lastTs           uint32
+	nans             int32   // the partition currently being checked - nope?
+	deltaSum         float64 // sum of abs(value - ts) across the time series
+	numNonMatching   int32   // number of timestamps where value != ts
+	lastSeen         uint32  // the last seen non-NaN time stamp (useful for lag)
+	correctNumPoints bool    // whether the expected number of points were received
+	correctAlignment bool    // whether the last ts matches `now`
+	correctSpacing   bool    // whether all points are sorted and 1 period apart
+}
+
+type partitionMetrics struct {
+	nanCount         *stats.Gauge32 // the number of missing values for each series
+	lag              *stats.Gauge32 // time since the last value was recorded
+	deltaSum         *stats.Gauge32 // the total amount of drift between expected value and actual values
+	nonMatching      *stats.Gauge32 // total number of entries where drift occurred
+	correctNumPoints *stats.Bool    // whether the expected number of points were received
+	correctAlignment *stats.Bool    // whether the last ts matches `now`
+	correctSpacing   *stats.Bool    // whether all points are sorted and 1 period apart
+}
+
+func monitor() {
+	initMetricsBySeries()
+	for tick := range clock.AlignedTickLossless(queryInterval) {
+
+		query := graphite.ExecuteRenderQuery(buildRequest(tick))
+		if query.HTTPErr != nil {
+			httpError.Inc()
+			continue
+		}
+		if query.DecodeErr != nil {
+			decodeError.Inc()
+			continue
+		}
+
+		for _, s := range query.Decoded {
+			processPartitionSeries(s, tick)
+		}
+	}
+}
+
+func processPartitionSeries(s graphite.Series, now time.Time) {
+	partition, err := strconv.Atoi(s.Target)
+	if err != nil {
+		log.Debug("unable to parse partition", err)
+		invalidError.Inc()
+		return
+	}
+	if len(s.Datapoints) < 2 {
+		log.Debugf("partition has invalid number of datapoints: %d", len(s.Datapoints))
+		invalidError.Inc()
+		return
+	}
+
+	serStats := seriesStats{}
+	serStats.lastTs = s.Datapoints[len(s.Datapoints)-1].Ts
+	serStats.correctAlignment = int64(serStats.lastTs) == now.Unix()
+	serStats.correctNumPoints = len(s.Datapoints) == int(lookbackPeriod/testMetricsInterval)+1
+	serStats.correctSpacing = checkSpacing(s.Datapoints)
+
+	for _, dp := range s.Datapoints {
+		if math.IsNaN(dp.Val) {
+			serStats.nans += 1
+			continue
+		}
+		serStats.lastSeen = dp.Ts
+		if diff := dp.Val - float64(dp.Ts); diff != 0 {
+			log.Debugf("partition=%d dp.Val=%f dp.Ts=%d diff=%f", partition, dp.Val, dp.Ts, diff)
+			serStats.deltaSum += diff
+			serStats.numNonMatching += 1
+		}
+	}
+
+	metrics := metricsBySeries[partition]
+	metrics.nanCount.Set(int(serStats.nans))
+	lag := lastPublish - int64(serStats.lastSeen)
+	metrics.lag.Set(int(lag))
+	metrics.deltaSum.Set(int(serStats.deltaSum))
+	metrics.nonMatching.Set(int(serStats.numNonMatching))
+	metrics.correctNumPoints.Set(serStats.correctNumPoints)
+	metrics.correctAlignment.Set(serStats.correctAlignment)
```

Comment:
you're repeatedly (re)setting the same bool for each partition checked.
so if the stats for 127 partitions are bad, but the last one is good, everything will still look good.
i recommend you use counters here, not bools.

Reply:
There's a metrics object per partition which we're accessing right above this (`metrics := metricsBySeries[partition]).

I'm calling `correctAligntment.Set(serStats.correctAlignment)` on 128 different stats objects.

(unless I'm missing somet)

Should I still use counters instead of bools here?